Implement mutation validation

fedepaol opened the pull request:
**What this PR does / why we need it**:

This adds validation for both Assign and AssignMeta crds.

**Which issue(s) this PR fixes** *(optional, using `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when the PR gets merged)*:
Fixes #977

**Special notes for your reviewer**:
Based on https://github.com/open-policy-agent/gatekeeper/pull/938 and https://github.com/open-policy-agent/gatekeeper/pull/881
Last commit contains validation.

maxsmythe commented:
Looks like the linter wants you to turn a string into a constant.

maxsmythe commented:
LMK when this is ready for review!

fedepaol commented:
> LMK when this is ready for review!

It's ready. I added WIP because of its dependencies, but https://github.com/open-policy-agent/gatekeeper/pull/978/commits/d2e6165ebde2b9a89f7891b6c7977619524bb165 is good for review!

maxsmythe commented on pkg/mutation/assign_mutator.go:
```diff
@@ -123,3 +127,38 @@ func applyToToBindings(applyTos []mutationsv1alpha1.ApplyTo) []SchemaBinding {
 	return res
 }
+
+// IsValidAssign returns an error if the given assignmetadata object is not
```
nit: `assignMetadata` -> `assign`

fedepaol replied:
Done

maxsmythe commented on pkg/mutation/assign_mutator.go:
```diff
@@ -123,3 +127,38 @@ func applyToToBindings(applyTos []mutationsv1alpha1.ApplyTo) []SchemaBinding {
+	return nil
+}
+
+func hasMetadataRoot(path *parser.Path) bool {
```
Do we want to wire in the other validation logic in #938 ?

fedepaol replied:
This is slightly different.
The other logic checks that the path is exactly metadata.labels.xxx or metadata.annotations.xxx because those are the only valid paths (so far).

Here, it's enough to have metadata as a root to make the assign object invalid.

maxsmythe replied:
Good point, SGTM

maxsmythe commented on pkg/mutation/assignmeta_mutator.go:
```diff
@@ -123,3 +124,29 @@ func isMetadataPath(path *parser.Path) bool {
+	if err != nil {
+		return errors.Wrapf(err, "Invalid location format")
+	}
+	if !isMetadataPath(path) {
```
nit: should this be `isValidMetadataPath()` to indicate that it's not just testing whether the path has a metadata root?

fedepaol replied:
Done

maxsmythe commented on pkg/webhook/mutation_test.go:
```diff
@@ -0,0 +1,232 @@
+)
+
+const (
+	validAssignMeta = `
```
Should we just embed these strings in the test if we aren't going to reuse them?

fedepaol replied:
Yep. I "took inspiration" from https://github.com/open-policy-agent/gatekeeper/blob/f3ec64a56154cd991e98387329083d243c967879/pkg%2Fwebhook%2Fpolicy_test.go but using this locally will make it clearer.

fedepaol replied:
Done

maxsmythe reviewed (commented):
This mostly LGTM, though it is blocked by the webhook PR. Worth getting other eyes on.

ritazh commented:
This PR looks pretty good pending few comments. Will review again after #881 is merged.

fedepaol commented:
> This PR looks pretty good pending few comments and maybe update `mutations` to `mutation`? Will review again after #881 is merged.

for the /s/mutations/mutation as we discussed in another pr, I'd wait until the end so we do it only once.

maxsmythe commented:
+1 to fixing cleanup things like mutations -> mutation at the end

fedepaol commented:
@ritazh @maxsmythe rebased against master

maxsmythe commented on pkg/webhook/mutation.go:
```diff
@@ -159,3 +174,43 @@ func AppendMutationWebhookIfEnabled(webhooks []rotator.WebhookInfo) []rotator.We
+func (h *mutationHandler) validateAssignMetadata(ctx context.Context, req admission.Request) (bool, error) {
+	obj, _, err := deserializer.Decode(req.AdmissionRequest.Object.Raw, nil, &mutationsv1alpha1.AssignMetadata{})
+	if err != nil {
+		return false, err
```
It might be user error if we can't deserialize the resource into the appropriate object.

Then again, that should have been caught by the API server checking the schema of the CRD, so maybe this is the appropriate way to handle things.

TL;DR I can see this both ways, so fine as-is

maxsmythe commented on pkg/webhook/mutation.go:
```diff
@@ -159,3 +174,43 @@ func AppendMutationWebhookIfEnabled(webhooks []rotator.WebhookInfo) []rotator.We
+	if err != nil {
+		return false, err
+	}
+	assignMetadata := obj.(*mutationsv1alpha1.AssignMetadata)
```
we should do a safe type casting here (e.g. `assignMetadata, ok := obj.(*mutationsv1alpha1.AssignMetadata)`)

maxsmythe commented on pkg/webhook/mutation.go:
```diff
@@ -159,3 +174,43 @@ func AppendMutationWebhookIfEnabled(webhooks []rotator.WebhookInfo) []rotator.We
+	if err != nil {
+		return false, err
+	}
+	assign := obj.(*mutationsv1alpha1.Assign)
```
safe type casting

maxsmythe commented on pkg/webhook/mutation_test.go:
```diff
@@ -0,0 +1,220 @@
+metadata:
+  name: goodAssign
+spec:
+  location: "spec.containers[name:test].foo"
```
can we add in a test that makes sure we are denying `spec.containers[name: test].test` here?

If that test doesn't currently pass, we can comment it out and add a TODO to uncomment it, I just don't want to forget it

fedepaol replied:
Lol basically the example I put as a good example is not a good one :-)

fedepaol replied:
Added the validation and the test

maxsmythe commented on pkg/webhook/mutation_test.go:
```diff
@@ -0,0 +1,220 @@
+			ErrorExpected: true,
+		},
+		{
+			Name: "No Assign",
```
Another test that we want to reject early is mutations where:

```
spec.containers[name: foo] = << anything other than an object >>
```

This can also likely be commented out for now and noted with a TODO, as I don't think we are checking for this case currently.

fedepaol replied:
Added the validation and the test

maxsmythe reviewed (changes requested):
Basically ready.

Some quick changes needed for safe type casting and some validation invariants that we'll want to assert in the future (even if we're not currently doing so now)

maxsmythe commented on pkg/mutation/assign_mutator.go:
```diff
@@ -131,3 +135,80 @@ func applyToToBindings(applyTos []mutationsv1alpha1.ApplyTo) []SchemaBinding {
+	lastNode := p.Nodes[len(p.Nodes)-1]
+	secondLastNode := p.Nodes[len(p.Nodes)-2]
+	// the format spec.containers[name:foo].bar is not allowed
+	if secondLastNode.Type() == parser.ListNode && lastNode.Type() == parser.ObjectNode {
```
The child of a list actually must be an object. What should be disallowed is:

`spec.containers[name: foo].name = bar`

where users are changing the key value.

We also want to check to make sure people aren't writing:

`spec.containers[name: foo] = {name: bar}`

which does the same thing.

fedepaol replied:
My bad, I totally misunderstood the ask.

maxsmythe replied:
No worries!

maxsmythe commented on pkg/mutation/assign_mutator.go:
```diff
@@ -131,3 +135,80 @@ func applyToToBindings(applyTos []mutationsv1alpha1.ApplyTo) []SchemaBinding {
+	if p.Nodes[len(p.Nodes)-1].Type() != parser.ListNode {
+		return nil
+	}
+	_, ok := value.(map[string]interface{})
```
+1 this is awesome!

The last step here is to validate that:

`valueMap[listNode.KeyField] == listNode.KeyValue`

(the field must both be present and equal to the expected value)

maxsmythe commented on pkg/webhook/mutation_test.go:
```diff
@@ -0,0 +1,251 @@
+    assign:
+      value: xxx
+`,
+			ErrorExpected: true,
```
This should be `ErrorExpected: false`

fedepaol commented:
@maxsmythe I think I fixed this, plus I added another check since we can't have the last node as a globbed list (I think), as we would add the same object to all the items (and the key can't be the same by definition).

maxsmythe commented on pkg/mutation/assign_mutator.go:
```diff
@@ -131,3 +135,110 @@ func applyToToBindings(applyTos []mutationsv1alpha1.ApplyTo) []SchemaBinding {
+	}
+
+	if hasMetadataRoot(path) {
+		return fmt.Errorf("Assign can't change metadata")
```
nit: `Errorf` -> `errors.New()`, since we arent using any formatting args

Also, golang convention is that error strings are lowercase:

https://github.com/golang/go/wiki/CodeReviewComments#error-strings

This is in a few places, but I'm only leaving the one comment

maxsmythe commented on pkg/mutation/assign_mutator.go:
```diff
@@ -131,3 +135,110 @@ func applyToToBindings(applyTos []mutationsv1alpha1.ApplyTo) []SchemaBinding {
+	toAssign := make(map[string]interface{})
+	err = json.Unmarshal([]byte(assign.Spec.Parameters.Assign.Raw), &toAssign)
+	if err != nil {
+		return errors.Wrapf(err, "Invalid format for parameters.assign")
```
nit: `Wrapf` -> `Wrap`, since we arent using formatting args

This is in a few places, but I'm only leaving the one comment

maxsmythe reviewed (approved):
Thanks for this!

You are 100% right about the globbed list thing.

This LGTM, I only have superficial nits.

@ritazh LGTY?

LMK when ready to submit

fedepaol commented:
I think I fixed the nits, this is good to go on my side.

ritazh commented on pkg/mutation/assign_mutator.go:
```diff
@@ -131,3 +135,110 @@ func applyToToBindings(applyTos []mutationsv1alpha1.ApplyTo) []SchemaBinding {
+
+// checkKeyNotChanged does not allow to change the key field of
+// a list element. A path like foo[name: bar].name is rejected
+func checkKeyNotChanged(p *parser.Path) error {
```
+1 for adding this

ritazh commented on pkg/webhook/mutation_test.go:
```diff
@@ -0,0 +1,299 @@
+	}
+}
+
+func TestAssignValidation(t *testing.T) {
```
Thanks for adding these tests!

ritazh reviewed (approved):
lgtm
